descartes_planner/src/planning_graph.cpp

```diff
@@ -36,16 +36,108 @@
 
+#include <future>
+
 using namespace descartes_core;
 using namespace descartes_trajectory;
+
+namespace
+{
+  using JointSolutions = std::vector<std::vector<double>>;
+  using VecJointSolutions = std::vector<JointSolutions>;
+  using ConstTrajectoryIterator = std::vector<descartes_core::TrajectoryPtPtr>::const_iterator;
+
+  using JointSolutionResult = std::pair<VecJointSolutions, bool>; // result and flag if everything went well
+
+  JointSolutionResult
+  rangeCalculateJointSolutions(descartes_core::RobotModelPtr model, 
+                               ConstTrajectoryIterator begin,
+                               ConstTrajectoryIterator end)
+  {
+    JointSolutionResult solutions;
+    solutions.first.reserve(std::distance(begin, end));
+    solutions.second = true;
+
+    while (begin != end)
+    {
+      JointSolutions joint_poses;
+      begin->get()->getJointPoses(*model, joint_poses);
+
+      if (joint_poses.empty())
+      {
+        solutions.second = false;
+        ROS_ERROR("No joint solutions available for point %s", 
+          boost::uuids::to_string(begin->get()->getID()).c_str());
+      }
+
+      solutions.first.push_back(std::move(joint_poses));
+      ++begin;
+    }
+
+    return solutions;
+  }
+
+  JointSolutionResult
+  parallelCalculateJointSolutions(const std::vector<descartes_core::TrajectoryPtPtr>& trajectory,
+                                  const descartes_core::RobotModel& model,
+                                  const unsigned max_threads)
+  {
+    // Might want to insert check to see if work to be done is small and if so,
+    // just run the IKs synchronously
+    JointSolutionResult solutions;
+    solutions.first.reserve(trajectory.size());
+    solutions.second = true; // success by default
+
+    // Divide work into chunks
+    const unsigned chunk_size = trajectory.size() / max_threads;
+    const unsigned chunk_extra = trajectory.size() % max_threads;
+
+    // Create tasks
+    std::vector<std::future<JointSolutionResult>> futures (max_threads);
+    for (size_t i = 0; i < futures.size(); ++i)
+    {
+      auto start = trajectory.cbegin() + i * chunk_size;
+      auto end = start + chunk_size;
+
+      // Give extra work to last element
+      if (i == futures.size() - 1) end += chunk_extra;
+
+      // Create a seperate copy of the robot state for this worker
+      descartes_core::RobotModelPtr state_copy = model.clone();
+      // Launch task
+      futures[i] = std::async(std::launch::async, rangeCalculateJointSolutions, state_copy, start, end);
+    }
+
+    // Wait for solutions
+    for (auto& future : futures)
+    {
+      JointSolutionResult solution_segment = std::move(future.get());
+
+      solutions.second = solutions.second && solution_segment.second; // carry the false flag forward if is returned
+      if (!solution_segment.second)
+      {
+        ROS_ERROR("One segment of the calculation failed to find a solution for all points");
```

Comment:
A bad point (or one without any IK solutions) results in (by definition) no plan.  The current MoveIt cartesian planner ignores bad points and returns a partial path.  We do NOT want to do this.  @jrgnicho and I had some discussion (in person) on this for PR #49.  If I am remembering correctly, we decided that these were the requirements for a planner (@jrgnicho feel free to correct me).
 * A planner should fail quickly if plan is not possible.  A proper return code should indicate why.
 * Additional, possibly slow methods should allow the user to get more information about the failure.
This approach balances responsiveness with user feedback.  One could imagine an application where knowing whether a plan is possible is all that is required.  Taking additional time to disect the failure, by default, is of no value to the user. However, another application such as a GUI, would want more detailed information about a failure so that the user could correct the problem points.